src/common/src/util/must_not_impl.rs

```diff
@@ -0,0 +1,17 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Magic trait to prevent `T` from implementing `Ord`.
+pub trait MustNotImplOrd {}
```

Comment:
How does the magic work?

Reply:
If some developer `impl Ord for ScalarImpl` by accident, there'll be a conflict.

Reply:
Hmm I'm not entirely convinced this works.
I prefer the more declarative `negative_impls` approach: https://github.com/kwannoel/rust-experiments/blob/main/ban-trait2.rs

I tried something like this approach here: https://github.com/kwannoel/rust-experiments/blob/main/ban-trait.rs but seems like it does not prevent it.

Reply:
> I tried something like this approach here: [kwannoel/rust-experiments@main/ban-trait.rs](https://github.com/kwannoel/rust-experiments/blob/main/ban-trait.rs?rgh-link-date=2023-07-27T03%3A51%3A47Z) but seems like it does not prevent it.

You have to `impl NotOrd for Custom {}` so that it'll report a conflict

Reply:
It seems that `!Ord` is just like a syntax sugar for the hand-made `MustNotImplOrd`

Reply:
Know about this unstable feature for the first time. 🤯

Reply:
Oh I see how this works now.

1. We implement a blanket instance for T: PartialOrd (This is part of the `MustNotImplOrd` feature actually).
2. If we opt in for any type, they cannot also implement `PartialOrd`, otherwise it will conflict with this implementation.
3. Hence by implementing `impl MustNotImplOrd for ScalarImpl`, that makes sure it can't implement `PartialOrd`, and `Ord` by extension.

This approach makes sense to me.

I slightly prefer using the unstable `negative_impls` because it's more declarative, and the error message would be clearer:
```
error[E0751]: found both positive and negative implementation of trait `Ord` for type `Custom`:
  --> ban-trait2.rs:10:1
   |
7  | impl !Ord for Custom {}
   | -------------------- negative implementation here
...
10 | impl Ord for Custom {}
   | ^^^^^^^^^^^^^^^^^^^ positive implementation here
```

I'm okay with this approach too, but we should add some docs on how it works, because it is not intuitive 🤣 .